README.md
# NAME

App::Uc::TwitterIrcGateway - Twitter IRC Gateway of me by me for me

# SYNOPSIS

## Configure consumer key that utig.pl uses

    $ script/utig.pl configure

    input Twitter consumer key:
    input Twitter consumer secret:

## Start twitter irc gateway server

    $ script/utig.pl run --host 127.0.0.1 --port 16668

## Login utig.pl server

- IRCクライアントで起動したサーバに適当な名前でログインする
- 暫くするとサーバーから OAuth 認証用の URL を渡されるので、それを開いて認証する
- IRCクライアントで

        /pin <pin code>
- そのうちTwitterストリームの読み込みが始まる

# DESCRIPTION

utig.pl は userstream の監視プログラムに毛が生えた程度のTwitter IRCゲートウェイサーバです

# INSTALLATION

## GitHub Checkout

    $ git clone git@github.com:UCormorant/utig.pl.git
    $ cd utig.pl/
    $ git submodule update --init
    $ cpanm --installdeps .

    # and run utig.pl

    $ perl script/utig.pl run

## CPAN Minus

__\*it doesn't work yet!\*__

    $ cpanm git@github.com:UCormorant/utig.pl.git

    # and run utig.pl
    $ utig run

# FEATURES

- UserStream を使用して閲覧するので発言が即座に流れてくるぞ！
- あんまり API を叩かない仕様だから他の Twitter 関連サービスと併用しても安心！
- コマンドと自前で作った TypableMap が快適な Twitter Life をサポートするぞ！
- Lists 対応。ただしリストに入れていてもフォローしてない人の発言は流れてこないぞ！
しかも自分が作ったリストしか見れないぞ！
- MySQLにログたくさんとるぞ！(いまうごいてないです)
- Follow, unfollow, direct message, block, list, account の操作？そんなもんねぇ！
(いつか対応予定です)
- 設定は `$HOME/.utig` にみんな入ってる

## ACTION COMMANDS

CTCP-actionにいろんなコマンドを実装してあります。

    /me *command* *args*

上記のような感じで使います。

    /me mention (or me): fetch mentions.
    /me reply (or re) <tid> <text>: reply to a <tid> tweet.
    /me favorite (or f, fav) +<tid>: add <tid> tweets to favorites.
    /me unfavorite (or unf, unfav) +<tid>: remove <tid> tweets from favorites.
    /me retweet (or rt) +<tid>: retweet <tid> tweets.
    /me quotetweet (or qt, qw) <tid> <text>: quotetweet a <tid> tweet, like "<text> QT \@tid_user: tid_tweet".
    /me delete (or del, oops) *<tid>: delete your <tid> tweets. if unset <tid>, delete your last tweet.
    /me list (or li) <screen_name>: list <screen_name>'s recent 20 tweets.
    /me information (or in, info) +<tid>: show <tid> tweets information. e.g. retweet_count, has conversation, created_at.
    /me conversation (or co) <tid>: show <tid> tweets conversation.
    /me ratelimit (or rate, limit): show remaining api hit counts.

だいたいこんな感じ！

    [SPACE] f <tid>...

みたいに、先頭に空白入れると action として認識するのでいちいち `/me ` とか打たなくてもへいきです

# DEPENDENCIES

## App::Uc::TwitterIrcGateway

- [perl](https://metacpan.org/pod/perl) >= 5.14
- Uc::IrcGateway [https://github.com/UCormorant/p5-uc-ircgateway](https://github.com/UCormorant/p5-uc-ircgateway)
- Uc::Model::Twitter [https://github.com/UCormorant/p5-uc-model-twitter](https://github.com/UCormorant/p5-uc-model-twitter)
- [Net::Twitter::Lite](https://metacpan.org/pod/Net::Twitter::Lite)
- [AnyEvent::Twitter](https://metacpan.org/pod/AnyEvent::Twitter)
- [AnyEvent::Twitter::Stream](https://metacpan.org/pod/AnyEvent::Twitter::Stream)
- [Clone](https://metacpan.org/pod/Clone)
- [Config::Pit](https://metacpan.org/pod/Config::Pit)
- [DateTime::Format::DateParse](https://metacpan.org/pod/DateTime::Format::DateParse)
- [HTML::Entities](https://metacpan.org/pod/HTML::Entities)
- [namespace::clean](https://metacpan.org/pod/namespace::clean)

## utig.pl

- [Data::Lock](https://metacpan.org/pod/Data::Lock)
- [Smart::Options](https://metacpan.org/pod/Smart::Options)
- [Term::ReadKey](https://metacpan.org/pod/Term::ReadKey)

# BUGS AND LIMITATIONS

Please report any bugs or feature requests to
[https://github.com/UCormorant/p5-app-uc-twitterircgateway/issues](https://github.com/UCormorant/p5-app-uc-twitterircgateway/issues)

# AUTHOR

[https://twitter.com/c18t](https://twitter.com/c18t)

# LICENCE AND COPYRIGHT

Copyright (C) U=Cormorant.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See [perlartistic](https://metacpan.org/pod/perlartistic).


Changes
Revision history for utig.pl

{{$NEXT}}

v1.1.5 2014-02-08T13:36:53Z

    - bugfix: tweet2db's remark2db doesn't work well

    - updated p5-uc-model-twitter: you should migrate DB
        - SQL files for migration are in
          ./extlib/p5-uc-model-twitter/migrate/v1.1.2_to_v1.2.0/

v1.1.4 2014-02-08T05:35:35Z

    - added features:
        - processing event stream: notice some event that is to you
        - TRIAL option 'in_memory':
            - all of connection's data store into memory.
              it will save as sqlite file when disconnecting.

    - bugfix: use SSL for that Twitter enforces SSL encryption for apps connecting to its API

v1.1.3 2014-01-01T07:54:39Z

    - bugfix: fatal error occered during vacuum tweet db.

v1.1.2 2014-01-01T00:54:01Z

    - performance improvement of get_tweet/set_tweet:

        - get_tweet speed is increased by 900%.
        - set_tweet speed is increased by 30%.
        - db file size thin to 1/20.

    - bugfix: display event user name to retweeted status user name.

    - add RT mark to head of RT messages.

    - add benchmark bench/component_tweet_gettter_and_setter.pl

v1.1.1 2013-12-31T01:20:59Z

    - update cpanfile:

        - add dependencies of extlib

    - update POD: INSTALLATION

v1.1.0 2013-12-30T22:05:18Z

    - add feature tweet2db

        - with option '--tweet2db', utig.pl collects streaming to MySQL.
          before using this option, you should command following:

            echo CREATE DATABASE twitter DEFAULT CHARACTER SET utf8mb4 | mysql -u root -p

    - bugfix: these DBI options are not enabled in Component/Tweet.pm:

        - RaiseError, PrintError, AutoCommit

    - bugfix: {user_id}.tweet.sqlite gets fat endlessly.

v1.0.1 2013-12-29T21:05:29Z

    - bugfix: fatal on twitter_configure when invalid consumer key is given.

    - utig.pl: add sub command 'configure' and 'run'

        - configure: update and verify consumer key that utig.pl uses.
        - run: run utig.pl(irc gateway server)

    - update documents

        - SYNOPSIS and INSTALLATION are update to v1.0.1

    - update cpanfile

v1.0.0 2013-12-28T23:44:11Z

    - version 1.0.0

    - come back features

        - action commands
        - typable map
        - get mention

    - bugfix: idle time is never reset.
    - bugfix: long retweet text is cut off its tail.

    - reject "geven ~ when".

v0.1.0 2013-12-27T20:30:14Z

    - return some function from parent project.

        - oauth, register, lists, streamer, process_tweet, post tweet

    - add features

        - userdata in sqlite

    - add attributes

        - consumer_key, consumer_secret

    - bugfix: login user's nick and login are broken on register.
    - bugfix: fetch_list grabs join command to spend much time.
    - bugfix: user and channel data are not updated.
    - bugfix: oauth authentication cannot be completed.
    - bugfix: no provide own tweet to the list includes oneself.

v0.0.0 2013-11-10T08:51:31Z

    - new project as a spin-off from Uc::IrcGateway



bench/component_tweet_gettter_and_setter.pl
#!/usr/bin/env perl

use 5.014;
use warnings;
use utf8;
use autodie;

use Benchmark qw(:all :hireswallclock);

use File::Basename qw(dirname basename);
use File::Spec::Functions qw(catdir catfile);
use lib catdir(dirname(__FILE__), '..', 'extlib', 'p5-uc-ircgateway', 'lib');
use lib catdir(dirname(__FILE__), '..', 'lib');
use lib catdir(dirname(__FILE__));
use lib::Util;

use Uc::IrcGateway::Common qw(to_json from_json);
use Config::Pit qw(pit_get);

my $conf_app  = pit_get('utig.pl', require =>{
    consumer_key    => 'your twitter consumer_key',
    consumer_secret => 'your twitter consumer_secret',
});
my $conf_user = +{};
twitter_agent($conf_app, $conf_user);

my $db = basename(__FILE__) =~ s/\.\w+$//r;
my $tweets = sample_stream({
    consumer_key    => $conf_app->{consumer_key},
    consumer_secret => $conf_app->{consumer_secret},
    token           => $conf_user->{token},
    token_secret    => $conf_user->{token_secret},
}, 20);

my $dir = dirname(__FILE__);
cmpthese(30 => +{
    store_all => store_all(catfile($dir, "store_all_$db.sqlite")),
    store_nes => store_nes(catfile($dir, "store_nes_$db.sqlite")),
    store_all_txn => store_all(catfile($dir, "store_all_$db.sqlite"), 'enable_txn'),
    store_nes_txn => store_nes(catfile($dir, "store_nes_$db.sqlite"), 'enable_txn'),
}, 'auto');

cmpthese(60 => +{
    fetch_all => fetch_all(catfile($dir, "store_all_$db.sqlite")),
    fetch_nes => fetch_nes(catfile($dir, "store_nes_$db.sqlite")),
}, 'auto');

exit;

sub store_all {
    my ($db, $enable_txn) = @_;
    my $dbh = setup_dbh('SQLite',$db) or die "dbh connect error";
    $dbh->do(q{DROP TABLE IF EXISTS tweet});
    $dbh->do(q{
        CREATE TABLE 'tweet' (
          'id'    int  NOT NULL,
          'tweet' text,

          PRIMARY KEY ('id')
        )
    });
    $dbh->do(q{VACUUM});
    my $insert = sub {
        my $item = shift;
        my $sth = $dbh->prepare(q{
            INSERT OR REPLACE
                INTO tweet (id, tweet) VALUES (?, ?);
        });
        $sth->execute(rand 2**16, to_json($item, pretty => 0));
    };
    return $enable_txn
    ? sub {
        $dbh->begin_work;
        $insert->($_) for @$tweets;
        $dbh->commit;
    }
    : sub {
#        $dbh->do(q{DELETE FROM tweet});
        $insert->($_) for @$tweets;
    };
}

sub store_nes {
    my ($db, $enable_txn) = @_;
    my $dbh = setup_dbh('SQLite',$db) or die "dbh connect error";
    $dbh->do(q{DROP TABLE IF EXISTS tweet});
    $dbh->do(q{
        CREATE TABLE 'tweet' (
          'id'    int  NOT NULL,
          'text'  text,
          'login' text,
          'nick'  text,

          PRIMARY KEY ('id')
        )
    });
    $dbh->do(q{VACUUM});
    my $insert = sub {
        my $item = shift;
        my $sth = $dbh->prepare(q{
            INSERT OR REPLACE
                INTO tweet (id, text, login, nick) VALUES (?, ?, ?, ?);
        });
        my @values = (
            rand 2**16,
            $item->{text},
            $item->{user}{id},
            $item->{user}{screen_name},
        );
        $sth->execute(@values);
    };

    return $enable_txn
    ? sub {
        $dbh->begin_work;
        $insert->($_) for @$tweets;
        $dbh->commit;
    }
    : sub {
#        $dbh->do(q{DELETE FROM tweet});
        $insert->($_) for @$tweets;
    };
}

sub fetch_all {
    my $dbh = setup_dbh('SQLite',+shift) or die "dbh connect error";
    my $sth = $dbh->prepare(q{SELECT id FROM tweet});
    $sth->execute();
    my $result = $sth->fetchall_arrayref;
    my $select = sub {
        my $item_id = shift;
        my $sth = $dbh->prepare(q{
            SELECT tweet FROM tweet
                WHERE id = ?
        });
        $sth->execute($item_id);
        my $result = $sth->fetchrow_arrayref;
        $result ? from_json($result->[0]) : undef;
    };
    return sub {
        $select->($_->[0]) for @$result;
    };
}

sub fetch_nes {
    my $dbh = setup_dbh('SQLite',+shift) or die "dbh connect error";
    my $sth = $dbh->prepare(q{SELECT id FROM tweet});
    $sth->execute();
    my $result = $sth->fetchall_arrayref;
    my $select = sub {
        my $item_id = shift;
        my $sth = $dbh->prepare(q{
            SELECT * FROM tweet
                WHERE id = ?
        });
        $sth->execute($item_id);
        $sth->fetchrow_hashref;
    };
    return sub {
        $select->($_->[0]) for @$result;
    };
}

__END__
streamer starts to read... connected.
collect 20 tweets... done.

Benchmark: timing 30 iterations of store_all, store_all_txn, store_nes, store_nes_txn...
 store_all: 103.75 wallclock secs ( 0.50 usr +  2.37 sys =  2.87 CPU) @ 10.45/s(n=30)
store_all_txn: 8.43446 wallclock secs ( 0.26 usr +  0.31 sys =  0.58 CPU) @ 51.99/s (n=30)
 store_nes: 92.277 wallclock secs ( 0.33 usr +  2.26 sys =  2.59 CPU) @ 11.58/s(n=30)
store_nes_txn: 6.27834 wallclock secs ( 0.06 usr +  0.22 sys =  0.28 CPU) @ 106.38/s (n=30)
            (warning: too few iterations for a reliable count)

                Rate     store_all     store_nes store_all_txn store_nes_txn
store_all     10.5/s            --          -10%          -80%          -90%
store_nes     11.6/s           11%            --          -78%          -89%
store_all_txn 52.0/s          397%          349%            --          -51%
store_nes_txn  106/s          918%          818%          105%            --

Benchmark: timing 60 iterations of fetch_all, fetch_nes...
 fetch_all: 53.8766 wallclock secs (31.87 usr + 20.45 sys = 52.32 CPU) @  1.15/s (n=60)
 fetch_nes: 41.9675 wallclock secs (22.20 usr + 18.63 sys = 40.83 CPU) @  1.47/s (n=60)

            Rate fetch_all fetch_nes
fetch_all 1.15/s        --      -22%
fetch_nes 1.47/s       28%        --
